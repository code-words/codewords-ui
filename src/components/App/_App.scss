.backdrop {
  content: "";
  background: $overlay-filter;
  position: absolute;
  top: 0px;
  height: 100%;
  width: 100%;
  display: flex;
  justify-content: center;
  align-items: center;
}

.btn-default {
  border-radius: 20px;
  display: flex;
  align-items: center;
  justify-content: center;
  font-size: 16px;
  color: white;
  cursor: pointer;
  background: $button-default;
  text-decoration: none;

  &:hover {
    background: $button-hover;
  }

  &:disabled {
    color: gray;
    cursor: default;
    font-size: 1em;
    font-weight: 400;
    background: $button-default;
  }
}