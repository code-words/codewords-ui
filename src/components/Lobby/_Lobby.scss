.Lobby {
  position: relative;
  border-radius: 15px;
  height: 25em;
  width: 35em;
  background: $button-hover;
  display: flex;
  flex-direction: column;
  padding: 1em;

  .Lobby-head {
    height: 25%;
    padding-left: 20px;
    box-sizing: border-box;
    display: flex;
    align-items: center;

    .fade {
      content: "";
      height: 1.5px;
      width: 60%;
      margin-left: 20px;
      background: linear-gradient(90deg, rgb(126, 126, 126) 0%, rgba(255, 255, 255, 0) 100%);
    }
  }

  .Lobby-sub {
    height: 20%;
    box-sizing: border-box;
    display: flex;
    flex-direction: column;
    align-items: center;
    justify-content: center;
    
    h3 {
      color: rgb(126, 126, 126);
    }

    p {
      margin-top: 1em;

      code {
        padding-left: 10px;

        i {
          padding-left: 10px;
        }

        &:hover {
          cursor: pointer;
          font-weight: 700;
        }
      }
    }
  }

  .continue-btn {
    align-self: flex-end;
    padding: 5px 10px;
  }
}

.roster {
  box-sizing: border-box;
  padding: 30px;
  height: 70%;
}

.roster {
  h4 {
    margin: 10px 0;
    font-size: 24px;
  }

  .agent {
    background: gray;
    padding: .5em 1.5em;
    margin: .2em;
  }

  .missing {
    background: none;
    border: dashed 2px $faint-gray;
  }

  p {
    font-size: 22px;
    font-weight: 900;
    letter-spacing: .25em;
  }
}

#dot1 {
  animation:blinkingText 1.19s 0s infinite;
}
#dot2 {
  animation:blinkingText 1.19s .2s infinite;
}
#dot3 {
  animation:blinkingText 1.19s .4s infinite;
}
#dot4 {
  animation:blinkingText 1.19s .6s infinite;
}
#dot5 {
  animation:blinkingText 1.19s .8s infinite;
}

@keyframes blinkingText{
    0%{     color: #fff;    }
    49%{    color: transparent; }
    50%{    color: transparent; }
    99%{    color:transparent;  }
    100%{   color: #fff;    }
}