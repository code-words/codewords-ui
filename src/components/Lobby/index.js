import React from 'react';
import { Link } from 'react-router-dom';
import { CopyToClipboard } from "react-copy-to-clipboard";
const shortid = require('shortid');

const Lobby = ({ players, inviteCode, isLobbyFull }) => {
  const renderJoinedPlayers = () => {
    let agents = [];
    let availKeys = [];

    for (let i = 0; i <= 3; i++) {
      if (players[i]) {
        agents.push(players[i].name)
      } else {
        agents.push(null);
        availKeys.push(i)
      }
    }

    return agents.map((a, index) => isAgent(a, index));
  };

  const isAgent = (agent, index) => {
    if (agent !== null) {
      return (
        <div className="agent" key={shortid.generate()}>
          <p key={shortid.generate()}>{agent.toUpperCase()}</p>
        </div>
      )
    } else {
      return (
        <div className="agent missing" key={shortid.generate()}>
          <p key={shortid.generate()}>{loadingDots()}</p>
        </div>
      )
    }
  }

  const loadingDots = () => {
    let dots = [];

    for (let i = 1; i < 6; i++ ) {
      dots.push(`dot${i}`);
    }

    return dots.map((d, ind) => 
      <span key={ind} id={d}>-</span>
    )
  }

  const renderCode = () => {
    return (
      <p>
        Invite Code:
        <CopyToClipboard text={inviteCode}>
          <code>
            {inviteCode}
            <i className="far fa-copy" />
          </code>
        </CopyToClipboard>
      </p>
    );
  }

  const message = players.length === 4 ? `Ready to start mission! Click CONTINUE...`
    : `Waiting for ${4 - players.length} more players`;
  
  const messageCls = players.length === 4 ? 'Active' : '';

  const button = !isLobbyFull ? null
    : <Link className="btn-default continue-btn" to="/game">Continue</Link>;
  
  return (
    <div className="backdrop">
      <section className="Lobby">
        <div className="Lobby-head">
          <h2>LOBBY</h2>
          <span className="fade"></span>
        </div>
        <div className="Lobby-sub">
          <h3 className={`status ${messageCls}`}> 
            {message}
          </h3>
          {inviteCode && renderCode()}
        </div>
        <section className="roster">
          <h4>Current Agents:</h4>
          {renderJoinedPlayers()}
        </section>
        {button}
      </section>
    </div>
  )
}

export default Lobby;