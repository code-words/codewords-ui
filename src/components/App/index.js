import React, { Component } from 'react';
import { Route, Switch } from 'react-router-dom';
import Header from '../Header';
import StartScreen from '../StartScreen';
import RuleList from '../RuleList';
import NewGame from '../NewGame';
import JoinGame from '../JoinGame';
import Lobby from '../Lobby';
import Main from '../Main';
import ErrorScreen from '../ErrorScreen';
import Cable from 'actioncable';
import ConfDialog from '../ConfDialog';

export class App extends Component {
	constructor() {
		super();
		this.state = {
			user: {},
			invite_code: '',
      playerRoster: [],
      hintLogs: [],
      cardData: [],
      cable: {},
      isLobbyFull: false,
      currentPlayerId: null,
      remainingAttempts: null,
      currentHint: {
        hintWord: '', 
        relatedCards: null
      },
      showDialog: false,
      confMsg: ''
    };
  }

  handleUserInit = result => {
    const { id, name, token } = result;
    localStorage.setItem("Token", token);
    const user = { id, name, token }
    if (result.invite_code) this.setState({ invite_code: result.invite_code });
    this.setState({ user }, () => this.createCable(token));
  };

  updateHintLogs = data => {
    let hintLogs = this.state.hintLogs;
    hintLogs.push(data);
    this.setState({ hintLogs });
  };

  setPlayer = data => {
    const { playerRoster } = data;
    this.setState({ playerRoster }, () => {
      if (this.state.playerRoster.length === 4) {
        this.setState({ isLobbyFull: true })
      }
    })
  };

  setGame = async data => {
    console.log('game data: ', data)
    const { cards, players, firstPlayerId } = data;
		const user = players.find(p => p.id === this.state.user.id);
		let cardData = cards;

		if (user.isIntel) {
			const res = await fetch(`http://localhost:3000/api/v1/intel?token=${this.state.user.token}`);
			cardData = await res.json();
			cardData = cardData.cards;
    }

    this.setState({
      playerRoster: players,
      currentPlayerId: firstPlayerId,
			cardData,
			user: { ...this.state.user, ...user }
    });
  };
  
  setHint = data => {
    const { hintWord, relatedCards, currentPlayerId } = data;
    const player = this.state.playerRoster
      .find(p => p.id === this.state.currentPlayerId);

    this.setState({
      currentPlayerId,
      hint: { hintWord, relatedCards },
    }, () => this.showConf(`${player.name} gave a hint: ${hintWord} (x${relatedCards})`));
  }

	dataSwitch = result => {
    const { type, data } = result;
    console.log(type,':' ,data)

		switch (type) {
			case 'player-joined':
				this.setPlayer(data);
				break;
			case 'game-setup':
				this.setGame(data);
				break;
      case 'hint-provided':
        console.log('HINT GIVEN')
        this.setHint(data)
				break;
      case 'board-update':
        const player = this.state.playerRoster
          .find(p => p.id === this.state.currentPlayerId);
				const { card, currentPlayerId, remainingAttempts } = data;
				let cardData = [ ...this.state.cardData ];
        const cardIdx = cardData.findIndex(i => i.id === card.id);

        cardData[cardIdx] = { ...cardData[cardIdx], ...card };
        
        this.setState({ cardData, currentPlayerId, remainingAttempts }, () => this.showConf(`${player.name} guessed ${cardData[cardIdx].word}!`));
        
        break;
      case 'game-over':
        this.showConf(`End of Game!  ${data.winningTeam.toUpperCase()} team wins!`)
        break;
      case 'illegal-action':
        this.showConf(data.error);
        break;
			default:
				console.log('ERROR in Switch');
		}
  };
  
  showConf = msg => {
    this.setState({ showDialog: true, confMsg: msg });
    setTimeout(() => this.setState({ showDialog: false }), 2000);
  }

	createCable = token => {
		if (this.state.user.token === token) {
			let connection = Cable.createConsumer(`ws://localhost:3000/cable/${token}`);

			this.cable = connection.subscriptions.create(
				{ channel: 'GameDataChannel' },
				{
					connected: () => console.log('connected'),
					disconnected: () => console.log('disconnected'),
					rejected: () => console.log('rejected'),
					received: res => this.dataSwitch(JSON.parse(res.message)),
					sendHint: hint => {
            console.log('Hint MF', hint)
            this.cable.perform('send_hint', hint)
          },
					sendGuess: guess => {
						console.log('HIT CABLE METHOD GUESS', guess);
						this.cable.perform('send_guess', guess);
					}
				}
			);
			this.setState({ cable: this.cable });
		}
	};

	render() {
    console.log('state: ', this.state);

    const dialog = this.state.showDialog ? <ConfDialog message={this.state.confMsg} /> : null;
      
		return (
      <div className="App">
        {dialog}
				<Header />
				<Switch>
					<Route exact path="/" component={StartScreen} />
					<Route exact path="/rules" component={RuleList} />
					<Route exact path="/new" render={
            /* istanbul ignore next */
            () => <NewGame handleUserInit={this.handleUserInit} />} />
					<Route exact path="/join" render={
            /* istanbul ignore next */
            () => <JoinGame handleUserInit={this.handleUserInit} />} />
					<Route
						exact
						path="/lobby"
						render={
              /* istanbul ignore next */
              () => (
							<Lobby
								players={this.state.playerRoster}
								inviteCode={this.state.invite_code}
								isLobbyFull={this.state.isLobbyFull}
							/>
						)}
					/>
					<Route
						exact
						path="/game"
						component={() => (
							<Main
								token={this.state.user.token}
								hintLogs={this.state.hintLogs}
                cardData={this.state.cardData}
                isIntel={this.state.user.isIntel}
								isActive={this.state.user.id === this.state.currentPlayerId}
								cable={this.state.cable}
								players={this.state.playerRoster}
								sendGuess={this.cable.sendGuess}
							/>
						)}
					/>
					<Route component={ErrorScreen} />
				</Switch>
			</div>
		);
	}
}

export default App;
