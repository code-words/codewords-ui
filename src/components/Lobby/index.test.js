import React from 'react';
import Lobby from '.';
import { MemoryRouter } from 'react-router';
import App from '../App';
import { mount } from 'enzyme';

describe('Lobby', () => {
  let wrapper;
  let instance;
  let mockPlayers;

  beforeEach(() => {
    mockPlayers = [
      { name: 'test1', id: 1 },
      { name: 'test2', id: 2 }
    ];
    wrapper = mount(
      <MemoryRouter initialEntries={['/lobby']}>
        <App>
          <Lobby players={mockPlayers} />
        </App>
      </MemoryRouter>
    );
    instance = wrapper.instance();
  })

  it('should match component snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('should match snapshot after a player joins', () => {
    expect(wrapper).toMatchSnapshot();
    expect(mockPlayers.length).toEqual(2);

    mockPlayers.push({name: 'test3', id: 3 });

    expect(wrapper).toMatchSnapshot();
    expect(mockPlayers.length).toEqual(3);
  })

})