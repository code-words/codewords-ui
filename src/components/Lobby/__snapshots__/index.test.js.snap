// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Lobby should match component snapshot 1`] = `ReactWrapper {}`;

exports[`Lobby should match snapshot after a player joins 1`] = `ReactWrapper {}`;

exports[`Lobby should match snapshot after a player joins 2`] = `ReactWrapper {}`;